### PR TITLE
Porting HDF5 recorder environment along with Pinocchio Solver

### PR DESCRIPTION
**Pull request by zephirefaith**
## Motivation and Context

Solves Issue #60 

Ports over environments used for collecting and replaying demonstrations on Stretch. Also ports in Pinocchio IK solver which fulfills immediate need for manipulation motions. Since SLAP use-case is pretty similar to SANER's I am assuming this is the best way to attack the problem.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: projects/collect_h5.py
Comment (cpaxton) [inline projects/collect_h5.py:20]: type hints

Comment (cpaxton) [reply projects/collect_h5.py:20]: might be good to document what these 2 are as well

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:41]: type hints/descriptions

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:36]: 1 sentece class description

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:56]: hints

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:63]: hints

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:77]: should return type be (pos, quat) like this, or are we using sophus?
cc @exhaustin 

for now best to be consistent with pybullet i guess

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:86]: i don't like "while True", should probably be a max numebr of iter?

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:95]: oh i see, its here

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:159]: hints + descriptions

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:199]: could you make this into a short test??

Comment (cpaxton) [reply src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:199]: like in test/.../motions

File: src/home_robot/home_robot/agent/motion/stretch.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/stretch.py:195]: else - ik_type not recongized runtime error

File: src/home_robot/home_robot/agent/motion/stretch.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/stretch.py:203]: hints please

File: src/home_robot/home_robot/agent/motion/stretch.py
Comment (cpaxton) [inline src/home_robot/home_robot/agent/motion/stretch.py:465]: qi: np.ndarray

File: src/home_robot/home_robot/utils/data_tools/writer.py
Comment (cpaxton) [inline src/home_robot/home_robot/utils/data_tools/writer.py:94]: hard to do types here right? maybe we dont need hints. but key = str

**Review by cpaxton — COMMENTED** [review]
a few minor documentation fixes + maybe we should copy your testing logic into the tests folder?

File: src/home_robot/home_robot/utils/data_tools/writer.py
Comment (zephirefaith) [reply src/home_robot/home_robot/utils/data_tools/writer.py:94]: Yeah the values range from str, Dict, Lists of ndarrays. I will add key types only since that is consistent.

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (zephirefaith) [reply src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:86]: What is "consistent with pybullet"?

File: tests/home_robot/motion/test_pinocchio_ik.py
Comment (cpaxton) [inline tests/home_robot/motion/test_pinocchio_ik.py:10]: we should make this an actual test, like the other IK test -

**Review by cpaxton — COMMENTED** [review]
just one last thing, just check the inputs + outputs (vs forward kinematics) for your test to make sure its really working

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (exhaustin) [reply src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:87]: [pos, quat] is the pybullet standard: https://usermanual.wiki/Document/pybullet20quickstart20guide.479068914/html#pf17

I think we should unify pose representations across the entire repo but I don't have a very good candidate right now. Sophus is nice but it needs pybind to install which is slightly annoying. Maybe we can wait for the Theseus Lie library to release. Anyhow I think it's okay to just use pos, quat for now.

**Comment by zephirefaith** [comment]
@exhaustin I see this branch failing the tests on CI. I looked at the error message and it is due to the inability to find `rospkg`. Where can I add that requirement so we don't see this superfluous failure?

**Review by cpaxton — APPROVED** [review]
lgtm once tests are passing

**Review by exhaustin — CHANGES_REQUESTED** [review]
Sorry for the late review. 

As the original author of the Pinocchio IK solver and CEM implementation, I'm happy to help make changes in those places, feel free to delegate stuff to me.

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (zephirefaith) [reply src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py:171]: Could you take this task? Along with making this solution more generic?

File: tests/home_robot/motion/test_pinocchio_ik.py
Comment (zephirefaith) [reply tests/home_robot/motion/test_pinocchio_ik.py:68]: My reasoning wasn't to test if it ended early but that only asserting for error to be < tolerances provided is too tight a test. Thus to check full functionality I added checking for "best effort" optimization by checking `last_iter >= max_iter`

**Comment by zephirefaith** [comment]
@exhaustin I am assuming the indicator_block in the tests is for visualization and debugging purposes? That's my assumption on the decision to replicate it for both `pin_robot` and `pb_robot` for checking their FK+IK consistency against each other.

The basic logic is to take your `test_ik()` from `test_ik.py` but run those poses against both `pin_robot` and `pb_robot`. Then compute error/distance b/w the resultant `pos, quat` from both solvers to check for consistency between the solvers.

File: tests/home_robot/motion/test_pinocchio_ik.py
Comment (exhaustin) [reply tests/home_robot/motion/test_pinocchio_ik.py:85]: Sorry I don't follow. Are you're saying that you know that the problem you're giving to the optimizer is impossible, so you expect the optimizer to run until max iters as a best effort?

File: src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py
Comment (exhaustin) [reply src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py]: I just made the changes and also updated the tests, let me know if they look good!

**Comment by exhaustin** [comment]
> @exhaustin I am assuming the indicator_block in the tests is for visualization and debugging purposes? That's my assumption on the decision to replicate it for both `pin_robot` and `pb_robot` for checking their FK+IK consistency against each other.
> 
> The basic logic is to take your `test_ik()` from `test_ik.py` but run those poses against both `pin_robot` and `pb_robot`. Then compute error/distance b/w the resultant `pos, quat` from both solvers to check for consistency between the solvers.
> 
> Also, if you could take up the task to factor out current optimization-based solver to `UnderactuatedIKSolver` and make the current implementation more generic then that would be super-helpful! Thanks.

- Yes the block is for visualization, a nice addition from Chris.
- I just factored out the IK optimizer, let me know if it looks good.
- Now that the tests have more shared components and are more complicated, it'd be nice to use some [pytest.fixtures](https://docs.pytest.org/en/6.2.x/fixture.html). I can do it later if you don't have time, lmk.

**Comment by zephirefaith** [comment]
@exhaustin Changes look good to me. I like factoring out the optimizer as its own solver-agnostic class. Thanks!

I fixed the conflict and moved out `pinocchio_ik_solver.py` to the new motion folder to respect the new organization. Once the test passes let's merge this?

**Review by exhaustin — APPROVED** [review]
Approving now.
@zephirefaith feel free to leav comments on my changes, or just merge if my changes look good.

**Comment by zephirefaith** [comment]
Thanks for helping make the tests nice and modular @exhaustin 

Changes look good and the tests pass which is in itself a good marker!

**Pull request merged**
